<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Asynchronous JavaScript</title>
</head>

<body>
    <h1>Asynchronous JavaScript</h1>
    <script>

        // const result = fetch('https://restcountries.eu/rest/v2/name/portugal')
        // console.log(result)

        /*
        -) Promise is an object that is used as a placeholder for future result of an asynchronous operation
        -)Fetch also return as promise object
        */


        // fetch('https://restcountries.eu/rest/v2/name/portugal')
        //     .then((response) => {
        //         console.log(response)
        //         return response.json()
        //     })
        //     .then(data => {
        //         console.log(data[0])
        //     })

        /*
        -) Whatever we return from .then it becomes the fulfilled value of next .then
        -) here we returned a promise.
        -) if in line 23 I return 'Hi' then in line 25 data will get 'Hi'.
        */

        /*
        -) Catching errors
        -) We can catch errors in two ways first is by providing an error fucntion in each then
        */

        // fetch('https://restcountries.eu/rest/v2/name/portugal')
        //     .then(response => response.json(),
        //         err => console.log(err))
        //     .then(data => console.log(data[0]),
        //         err => console.log(err))

        /*-) Or by adding catch*/
        // fetch('https://restcountries.eu/rest/v2/name/portugal')
        //     .then(response => response.json())
        //     .then(data => console.log(data[0]))
        //     .catch(error => console.error(error))


        /*
        -) We have a finally (same as then and catch). 
        -) It will run at the end of chain always( no matter if the promise is resolved or rejected).
        -) Finally works on the value returned by catch
        */
        // fetch('https://restcountries.eu/rest/v2/name/portugal')
        //     .then(response => response.json())
        //     .then(data => console.log(data[0]))
        //     .catch(error => console.error(error))
        //     .finally(message => console.log(message))

        /*-) =We can create a manumal error by 'throw new Error()' if a condition satisfies*/

        /*-) To create a promise that resolves instanteneously we use ->*/

        // Promise.resolve('Resolved Promise')
        //     .then(res => console.log('hi'))

        /*
        -) .then of a promise is in MicroTask queue and other asynchronous tasks like setTimeout() are in callback queue
        -) Microtask have more priority than callback queue.
        -) So if we have a promise and timmer taking same time to complete ,
        -) promise will be executed first (irrespective of the timmers time, which may delay the timer)
        */

        /*
        -) Creating a Promise
        -) we are taking an example of lottery ticket system
        -) Added a timeout to simulate a real life promise
        */

        // const lotteryPromise = new Promise((resolve, reject) => {
        //     console.log('lotttery draw is happening')
        //     setTimeout(() => {
        //         if (Math.random() >= 0.5) {
        //             resolve('You win !!!')
        //         } else {
        //             reject('You Lost your money :(')
        //         }
        //     }, 2000)
        // })
        // lotteryPromise
        //     .then(res => console.log(res))
        //     .catch(err => console.log(err))


        /*
        -) Async and Await
        -) Async/Await provides a cleaner way to use Promises
        -) A function defined async means that the function will get paused when the async task is running in background.
        -) But the overvall execution will continue
        -) We use await before the async funtion to show that this function should wait for the promise to return a value
        -) Async funtion always return a promisef
        */

        const whereAmI = async (country) => {
            const response = await fetch(`https://restcountries.eu/rest/v2/name/${country}`)
            const data = await response.json()
            console.log(data)
            return `You are in ${country}`
            /*this is same as
            fetch('https://restcountries.eu/rest/v2/name/portugal')
                .then(response => response.json())
                .then(data => console.log(data[0]))
                */
        }
        const c = whereAmI('india')
        c.then(res => console.log(res)) // as we know async return a promise

    </script>
</body>

</html>}